packages/dataviews/src/components/dataform/stories/fixtures.tsx

```
@@ -0,0 +1,111 @@
+/**
+ * WordPress dependencies
+ */
+import {
+	ValidatedTextControl,
+	ValidatedTextareaControl,
+	ValidatedNumberControl,
+	ValidatedSelectControl,
+	ValidatedInputControl,
+} from '@automattic/components';
+
+/**
+ * Internal dependencies
+ */
+import type { Field, Form } from '../../../types';
+
+export type FormData = {
+	name: string;
+	surname: string;
+	email: string;
+};
+
+export const initialData: FormData = {
+	name: 'John',
+	surname: 'Doe',
+	email: 'john.doe@example.com',
+};
+
+export const getFields = (): Field< FormData >[] => [
+	{
+		id: 'name',
+		label: 'First Name',
+		type: 'text',
+		Edit: ( { field, data, onChange } ) => (
+			<ValidatedTextControl
+				label={ field.label }
+				value={ field.getValue( { item: data } ) }
+				onChange={ ( value: string ) =>
+					onChange( { [ field.id ]: value } )
+				}
+				customValidator={ () => {
+					return field?.isValid?.( data )
```

You mentioned:

> As a result, isValid returns a string with a default error message when using validation rules (like isRequired), instead when it is a custom validation function, it returns a boolean indicating validation status.

and 

> It would be interesting to explore an approach that, based on the isValid definition, field.isValid returns the right type. I will wait for your feedback before enhancing the type system and adding unit/integration tests.

Can you expand on that?

Based on what I saw so far, one option we have is the following:

- `field.isValid` always returns a boolean.
- Introduce a `field.getValidationErrors` that returns an array of messages for all the rules the field breaks. This is passed to the UI component to render them. This is assuming the current validation mechanism doesn't stop at the first error (I haven't checked how that works yet). I'm not opinionated on this, and it'd be fine to return a string for now if that's all that we need.

At the form level, I don't know that we need to return the error messages because we should handle that internally. So,`isItemValid` can keep as it is (returns a boolean).

---

Originally, I thought `isValid` would return a boolean for custom callbacks, and an error message or empty string otherwise.  
That said, I agree with your approach.`isValid` should consistently return a boolean. We can introduce a separate `getValidationErrors` utility to handle error messages.